### ISSUE TITLE
HealthCheck Failure results in UnhandledPromiseRejectionWarning: ServiceUnavailableException: Service Unavailable Exception

### ISSUE THREAD

**Issue by Scott8586** [mKeRix/room-assistant#923]
**Describe the bug**
Sometimes a HealthCheck Failure results in UnhandledPromiseRejectionWarning: ServiceUnavailableException: Service Unavailable Exception

**To reproduce**

Run a  bluetooth-low-energy integration, in the absence of btle devices.

**Relevant logs**
Log output that correlates to the bug. You can get debug output with `room-assistant -v`.

```text
Sep 26 16:43:58 raspi7 room-assistant[21446]: 26/09/2021, 16:43:58 - warn - BluetoothService: Did not detect any low energy advertisements in a while, resetting
Sep 26 16:44:13 raspi7 room-assistant[21446]: 26/09/2021, 16:44:13 - error - undefined: Health Check has failed! {"ble_advertisement_timeout":{"status":"down"}}
Sep 26 16:44:13 raspi7 room-assistant[21446]: (node:21446) UnhandledPromiseRejectionWarning: ServiceUnavailableException: Service Unavailable Exception
Sep 26 16:44:13 raspi7 room-assistant[21446]:     at HealthCheckService.<anonymous> (/opt/nodejs/lib/node_modules/room-assistant/node_modules/@nestjs/terminus/dist/health-check/health-check.service.js:63:19)
Sep 26 16:44:13 raspi7 room-assistant[21446]:     at Generator.next (<anonymous>)
Sep 26 16:44:13 raspi7 room-assistant[21446]:     at fulfilled (/opt/nodejs/lib/node_modules/room-assistant/node_modules/@nestjs/terminus/dist/health-check/health-check.service.js:14:58)
Sep 26 16:44:13 raspi7 room-assistant[21446]: (Use `node --trace-warnings ...` to show where the warning was created)
Sep 26 16:44:13 raspi7 room-assistant[21446]: (node:21446) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
Sep 26 16:44:13 raspi7 room-assistant[21446]: (node:21446) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Sep 26 16:44:13 raspi7 room-assistant[21446]: 26/09/2021, 16:44:13 - warn - BluetoothService: Did not detect any low energy advertisements in a while, resetting
Sep 26 16:44:28 raspi7 room-assistant[21446]: 26/09/2021, 16:44:28 - warn - BluetoothService: Did not detect any low energy advertisements in a while, resetting
Sep 26 16:44:43 raspi7 room-assistant[21446]: 26/09/2021, 16:44:43 - error - undefined: Health Check has failed! {"ble_advertisement_timeout":{"status":"down"}}
Sep 26 16:44:43 raspi7 room-assistant[21446]: (node:21446) UnhandledPromiseRejectionWarning: ServiceUnavailableException: Service Unavailable Exception
Sep 26 16:44:43 raspi7 room-assistant[21446]:     at HealthCheckService.<anonymous> (/opt/nodejs/lib/node_modules/room-assistant/node_modules/@nestjs/terminus/dist/health-check/health-check.service.js:63:19)
Sep 26 16:44:43 raspi7 room-assistant[21446]:     at Generator.next (<anonymous>)
Sep 26 16:44:43 raspi7 room-assistant[21446]:     at fulfilled (/opt/nodejs/lib/node_modules/room-assistant/node_modules/@nestjs/terminus/dist/health-check/health-check.service.js:14:58)
Sep 26 16:44:43 raspi7 room-assistant[21446]: (node:21446) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 4)
Sep 26 16:44:43 raspi7 room-assistant[21446]: 26/09/2021, 16:44:43 - warn - BluetoothService: Did not detect any low energy advertisements in a while, resetting
Sep 26 16:44:43 raspi7 systemd[1]: room-assistant.service: Watchdog timeout (limit 1min)!
Sep 26 16:44:43 raspi7 systemd[1]: room-assistant.service: Killing process 21446 (node) with signal SIGABRT.
Sep 26 16:44:43 raspi7 systemd[1]: room-assistant.service: Killing process 21536 (sh) with signal SIGABRT.
Sep 26 16:44:43 raspi7 systemd[1]: room-assistant.service: Killing process 21537 (hciconfig) with signal SIGABRT.
Sep 26 16:44:43 raspi7 room-assistant[21446]: 26/09/2021, 16:44:43 - error - undefined: Command failed: hciconfig hci0 reset
Sep 26 16:44:43 raspi7 systemd[1]: room-assistant.service: Main process exited, code=killed, status=6/ABRT
Sep 26 16:44:43 raspi7 systemd[1]: room-assistant.service: Failed with result 'watchdog'.
Sep 26 16:44:53 raspi7 systemd[1]: room-assistant.service: Service RestartSec=10s expired, scheduling restart.
Sep 26 16:44:53 raspi7 systemd[1]: room-assistant.service: Scheduled restart job, restart counter is at 1.
Sep 26 16:44:53 raspi7 systemd[1]: Stopped room-assistant service.
Sep 26 16:44:53 raspi7 systemd[1]: Starting room-assistant service...
```

**Relevant configuration**
Paste the relevant parts of your configuration below.

```yaml
global:
  instanceName: garage-vs
  integrations:
    - homeAssistant
    - bluetoothLowEnergy
homeAssistant:
  mqttUrl: 'mqtt://...."
  mqttOptions:
    username: xxx
    password: xxx
bluetoothLowEnergy:
  hciDeviceId: 0
  instanceBeaconEnabled: false
  onlyIBeacon: true
  updateFrequency: 15
  timeout: 90
  allowlist:
    - 'xxx-yyy-zzz'
```

**Expected behavior**

This should not result in an unhandled exception/crash.

**Environment**
 - room-assistant version: 2.19.0
 - installation type: npm7
 - hardware: Raspberry Pi Zero W
 - OS: Linux

**Additional context**

This exception does not occur when testing at the command line with '-v'.

One of my installs is in a rural environment, sometimes there are no btle devices around.  In addition to reporting the unhandled exception. I would like an option to disable this health check and/or make this situation not result in an error.

**Comment by Scott8586** [comment]
Apparently, this is sensitive to starting conditions, it does not occur on every command line or systemd based start.

**Comment by PeteBa** [comment]
Scott, I believe the log shows expected operation of room-assistant and not bugs or crashes.

The lines associated with `UnhandledPromiseRejectionWarning` are a reminder from the author of node that room-assistant might need updating before adopting a newer version as some things have changed from node 15 onward. This is just a warning and can be ignored if you are using node 14.

The lines with `Did not detect any low energy advertisements in a while, resetting` are, as you say, due to the lack of ble devices in the area. RA cant tell if the situation is due to a remote location or a hung ble query so performs a "safety" reset in an attempt to kick-start the ble chip. Again, expected behaviour.

The lines with `Health Check has failed!` are similar to the above but the next level of seriousness. If RA believes it is "unhealthy" (i.e. not received ble messages for a long time or mqtt connection is down etc), then it will trigger Debian's `systemd watchdog` service to do a full shutdown/restart of the application. The lines in the log with systemd in them are reflecting this restart activitty. Note that watchdog functionality is only enabled if running under systemd and not directly from the command line.

Can you post your room-assistant.service file ? You should be able to set `WatchdogSec=0` or just remove that line to turn off the watchdog stop/start functionality.

**Comment by Scott8586** [comment]
@PeteBa - Thanks again for a detailed description!  I jumped the gun in relation to the Exception - I read it code tripping over an unhandled exception, dying, and being resurrected by `systemd`.  I read up on `systemd` and now better understand the interaction of the program and the `watchdog` service.  I've disabled `WatchdogSec` (by commenting it out), and that keeps `systemd` from restarting the room-assistant service.  I'll call this closed.  Thanks again!
